### PR TITLE
feat(workspace): smart note refs

### PR DESCRIPTION
**Pull request by kevinslin**
- introduce smart note references as described in [[smart header references|dendron://dendron.docs/rfc.47-better-note-references#smart-header-references]]
- introduce `enableSmartRef` config option with default set to true
    - NOTE: decided to go with this instead of `enableLegacyRef` - the plan is to manually migrate everyone to the new syntax and remove legacy ref syntax
- fix bug in `extractHeader` (was mutating the block when extracting)
- copy note ref behavior update, if using `enableSmartRef` and a header is selected, don't generate `#{header},1:#*` but `#{header}` instead

Docs: https://github.com/dendronhq/dendron-site/pull/573

### CODE REVIEW

File: packages/common-all/data/dendron-yml.validator.json [hunk 1]
```diff
@@ -753,6 +753,9 @@
         "enableEditorDecorations": {
           "type": "boolean"
         },
+        "enableSmartRefs": {
```
Comment (jonathanyeung) [inline packages/common-all/data/dendron-yml.validator.json:756]: I'm not as keen on this name as I don't think it represents what the setting is doing.  I don't think one ref is 'smarter' or 'dumber' than the other, it's primarily just a different syntax.  So my preference would be like 'EnableHeaderGroupedRefSyntax' / 'EnableLegacyRefSyntax' (for the inverse setting) or something along those lines.

File: packages/plugin-core/package.json [hunk 2]
```diff
@@ -225,6 +225,10 @@
         "command": "dendron.randomNote",
         "title": "Dendron: Random Note"
       },
+      {
```
Comment (jonathanyeung) [inline packages/plugin-core/package.json:228]: unrelated to this review, but I think we don't actually need this command to be registered? Can we remove from const.ts (maybe separate review).

**Review by jonathanyeung — APPROVED** [review]
just one comment on the setting name, otherwise lgtm

File: packages/common-all/data/dendron-yml.validator.json (hunk 1 above)
Comment (kevinslin) [reply packages/common-all/data/dendron-yml.validator.json:756]: i can change it but i'm also planning on removing this as an option all together (we'll remove this in two weeks). its only an option to help with migration. do you think its worth the effort to change?

Comment (jonathanyeung) [reply packages/common-all/data/dendron-yml.validator.json:756]: nah- that's the best case :) let's leave it then

**Pull request merged**